flink-optimizer/src/main/java/org/apache/flink/optimizer/traversals/RangePartitionRewriter.java

```diff
@@ -0,0 +1,194 @@
+		final MapPartitionOperatorBase rbOperatorBase = new MapPartitionOperatorBase(rangeBoundaryBuilder, rbOperatorInformation, "RangeBoundaryBuilder");
+		final MapPartitionNode rbNode = new MapPartitionNode(rbOperatorBase);
+		final Channel rbChannel = new Channel(sicPlanNode, TempMode.NONE);
+		rbChannel.setShipStrategy(ShipStrategyType.FORWARD, channel.getDataExchangeMode());
+		final SingleInputPlanNode rbPlanNode = new SingleInputPlanNode(rbNode, "RangeBoundary PlanNode", rbChannel, DriverStrategy.MAP_PARTITION);
+		rbPlanNode.setParallelism(1);
+		rbChannel.setTarget(rbPlanNode);
+		sicPlanNode.addOutgoingChannel(rbChannel);
+		this.plan.getAllNodes().add(rbPlanNode);
+
+		// 4. Take range boundaries as broadcast input and take the tuple of partition id and record as output.
+		final AssignRangeIndex assignRangeIndex = new AssignRangeIndex(comparator);
+		final TypeInformation<Tuple2> ariOutputTypeInformation = new TupleTypeInfo<>(BasicTypeInfo.INT_TYPE_INFO, sourceOutputType);
+		final UnaryOperatorInformation ariOperatorInformation = new UnaryOperatorInformation(sourceOutputType, ariOutputTypeInformation);
+		final MapPartitionOperatorBase ariOperatorBase = new MapPartitionOperatorBase(assignRangeIndex, ariOperatorInformation, "Assign Range Index");
+		final MapPartitionNode ariNode = new MapPartitionNode(ariOperatorBase);
+		final Channel ariChannel = new Channel(sourceNode, TempMode.NONE);
+		// To avoid deadlock, set the DataExchangeMode of channel between source node and this to Batch.
+		ariChannel.setShipStrategy(ShipStrategyType.FORWARD, DataExchangeMode.BATCH);
+		final SingleInputPlanNode ariPlanNode = new SingleInputPlanNode(ariNode, "AssignRangeIndex PlanNode", ariChannel, DriverStrategy.MAP_PARTITION);
+		ariPlanNode.setParallelism(sourceParallelism);
+		ariChannel.setTarget(ariPlanNode);
+		this.plan.getAllNodes().add(ariPlanNode);
+		sourceNewOutputChannels.add(ariChannel);
+
+		final NamedChannel broadcastChannel = new NamedChannel("RangeBoundaries", rbPlanNode);
+		broadcastChannel.setShipStrategy(ShipStrategyType.BROADCAST, channel.getDataExchangeMode());
+		broadcastChannel.setTarget(ariPlanNode);
+		List<NamedChannel> broadcastChannels = new ArrayList<>(1);
+		broadcastChannels.add(broadcastChannel);
+		ariPlanNode.setBroadcastInputs(broadcastChannels);
+
+		// 5. Remove the partition id.
+		final Channel partChannel = new Channel(ariPlanNode, channel.getTempMode());
+		partChannel.setDataExchangeMode(channel.getDataExchangeMode());
+		final FieldList keys = new FieldList(0);
+		final boolean[] sortDirection = { true };
+		partChannel.setShipStrategy(channel.getShipStrategy(), keys, sortDirection, channel.getPartitioner(), channel.getDataExchangeMode());
+		ariPlanNode.addOutgoingChannel(channel);
+		partChannel.setLocalStrategy(channel.getLocalStrategy(), keys, sortDirection);
+		this.plan.getAllNodes().remove(targetNode);
```

Comment:
`Channel.setLocalStrategy()` happens during the period of `OptimizedNode.getAlternativePlans()` which is before the `RangePartitionRewriter` optimization,  so we have to set `LocalStrategy` even while keep the partition node and reuse the previous channel, as the key information has changed already. It seems removing partition node works well now, do you have some good reason why we should not implement in this way?